lib/Parse/ParseExpr.cpp

```diff
@@ -3007,6 +3016,107 @@ bool Parser::ParseRelativeBoundsClause(ExprResult &Expr) {
   return IsError;
 }
 
+ExprResult Parser::ParseBoundsCastExpression(IdentifierInfo &Ident,
+                                             SourceLocation &ILoc) {
+  ExprResult Result(true);
+  SourceLocation StartLoc = Tok.getLocation();
+  ParsedType CastTy = nullptr;
+  BoundsCastExpr::Kind kind;
+  if (&Ident == Ident_dynamic_bounds_cast)
+    kind = BoundsCastExpr::Kind::Dynamic;
+  else if (&Ident == Ident_assume_bounds_cast)
+    kind = BoundsCastExpr::Kind::Assume;
+  else
+    kind = BoundsCastExpr::Kind::Invalid;
+
+  DeclSpec DS(AttrFactory);
+  DS.SetRangeStart(Tok.getLocation());
+  DS.SetRangeEnd(SourceLocation());
+
+  if (ExpectAndConsume(tok::less, diag::err_expected_less_after,
+                       Ident.getNameStart()))
+    return ExprError();
+
+  TypeResult Ty = ParseTypeName();
+  if (Ty.isInvalid())
+    SkipUntil(tok::greater, StopAtSemi);
+
+  // if there is rel, parsing it.
+  // if(Tok.is(tok::comma)){
+  //   ConsumeToken();
+  //   Parsing Alignment
+  //   .....
+  // }
+
+  Declarator DeclaratorInfo(DS, Declarator::TypeNameContext);
+  ParseDeclarator(DeclaratorInfo);
+
+  SourceLocation EndLoc = Tok.getLocation();
+  if (ExpectAndConsume(tok::greater))
+    return ExprError(Diag(EndLoc, diag::note_matching) << tok::less);
+
+  DS.SetRangeEnd(EndLoc);
+
+  SourceLocation LParenLoc, RParenLoc;
+  BalancedDelimiterTracker T(*this, tok::l_paren);
+
+  if (T.expectAndConsume(diag::err_expected_lparen_after, Ident.getNameStart()))
+    return ExprError();
+  LParenLoc = T.getOpenLocation();
+
+  // Parsing e1, e2, e3
+  ExprResult CastExpr(true), CountExpr(true), RangeExpr(true);
```

Comment:
I think it would be clearer to use E1, E2, and E3 for the variable names.   I think CastExpr, CountExpr, and RangeExpr could confuse people later.